src/templateAccessibilityLabelForRule.ts

```diff
@@ -0,0 +1,107 @@
+import { BoundDirectivePropertyAst, ElementAst } from '@angular/compiler';
+import { sprintf } from 'sprintf-js';
+import { IRuleMetadata, RuleFailure, Rules, Utils } from 'tslint/lib';
+import { SourceFile } from 'typescript/lib/typescript';
+import { NgWalker } from './angular/ngWalker';
+import { BasicTemplateAstVisitor } from './angular/templates/basicTemplateAstVisitor';
+import { mayContainChildComponent } from './util/mayContainChildComponent';
+
+interface ILabelForOptions {
+  depth: number;
+  labelComponents: string[];
+  labelAttributes: string[];
+  controlComponents: string[];
+}
+export class Rule extends Rules.AbstractRule {
+  static readonly metadata: IRuleMetadata = {
+    description: 'Checks if the label has associated for attribute or a form element',
+    optionExamples: [
+      [true, { depth: 3, labelComponents: ['app-label'], labelAttributes: ['id'], controlComponents: ['app-input', 'app-select'] }]
+    ],
+    options: {
+      items: {
+        type: 'object',
+        properties: {
+          depth: {
+            type: 'number'
+          },
+          labelComponents: {
+            type: 'array',
+            items: {
+              type: 'string'
+            }
+          },
+          labelAttributes: {
+            type: 'array',
+            items: {
+              type: 'string'
+            }
+          },
+          controlComponents: {
+            type: 'array',
+            items: {
+              type: 'string'
+            }
+          }
+        }
+      },
+      type: 'array'
+    },
+    optionsDescription: 'Determine the depth of nesting for form control, custom label, label attribute and controls',
+    rationale: Utils.dedent`
+    The label tag should either have a for attribute or should have associated control.
+    This rule supports two ways, either the label component should explicitly have a for attribute or a control nested inside the label component
+    It also supports adding custom control component and custom label component support.`,
+    ruleName: 'template-accessibility-label-for',
+    type: 'functionality',
+    typescriptOnly: true
+  };
+
+  static readonly FAILURE_STRING = 'The Label element should have a for attribute or associating input inside label';
```

Comment:
Since the rule allows user to define label components and label attributes, IMHO we have to let it more generic.